app/models/minion.rb

```diff
@@ -9,37 +9,69 @@ class Minion < ApplicationRecord
   class NotEnoughMinions < StandardError; end
   # Raised when we fail to assign a role on a minion
   class CouldNotAssignRole < StandardError; end
+  # Raised when Minion doesn't exist
+  class NonExistingMinion < StandardError; end
 
   enum highstate: [:not_applied, :pending, :failed, :applied]
   enum role: [:master, :minion]
 
-  validates :hostname, uniqueness: true
+  validates :hostname, presence: true, uniqueness: true
 
-  # This method is used to assign the specified roles to Minions which do not
-  # have a roles already assigned (available minions).
-  # roles param are the roles we absolutely have to assign. If we can't assign
-  # one of those, the method will raise.
-  # default_role param can be set if we want all the rest of the available
-  # minions to get a default role.
-  # Returns the ids of the minions on which a roles has been assigned.
-  def self.assign_roles(roles: [], default_role: nil)
-    minions = Minion.where(role: nil)
-
-    # only the needed number of minions or all if we have a default role
-    minions = minions.limit(roles.size) unless default_role
-
-    raise NotEnoughMinions if minions.count < roles.size
+  # Example:
+  #   Minion.assign_roles(
+  #     roles: {
+  #       master: ["master.example.com"],
+  #       minion: ["minion1.example.com"]
+  #     },
+  #     default_role: :dns
+  #   )
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
```

Comment:
For example, this method is used like: "I want this role to be applied for this node, and I don't care about the rest (take a default)". Why not do this simpler:

- A method that assigns "expected" roles (I think that you don't even need to create it 😋)
- A method that sweeps all the marked nodes and effectively does the assignment in the kubernetes/salt context.

This two-step mechanism would give you simpler methods, while giving you the same functionality. Other questions like "finding a non-existing" minion would be handled by the callee, and if you want it can return an exception that would be handled as a mere 404 as I said before 😉